### PR TITLE
ServiceUndefine/DeleteInterface implementation and TestCase.

### PR DESCRIPTION
**Pull request by the-real-wizard-of-oz**

### CODE REVIEW

File: src/rest/mormot.rest.server.pas [hunk 1]
```diff
@@ -3019,65 +3023,74 @@ procedure TRestServerUriContext.ExecuteSoaByMethod;
 var
   timstart, timstop: Int64;
   sessionstat: TSynMonitorInputOutput;
+  method: PRestServerMethod;
```
Comment (synopse) [inline src/rest/mormot.rest.server.pas:3026]: you are right it is easier and safer to not use with ... but a local variable

File: src/core/mormot.core.interfaces.pas [hunk 2]
```diff
@@ -983,6 +983,7 @@   TInterfaceResolverInjected = class(TInterfaceResolver)
       overload; virtual;
     /// check if a given interface can be resolved, from its RTTI
     function Implements(aInterface: PRttiInfo): boolean; override;
+    procedure DeleteResolver(aResolver: TInterfaceResolver);
```
Comment (synopse) [inline src/core/mormot.core.interfaces.pas:986]: please add a /// comment (with 3 /) before the method for documentation consistency

File: src/core/mormot.core.interfaces.pas [hunk 3]
```diff
@@ -5391,6 +5392,14 @@ destructor TInterfaceResolverInjected.Destroy;
   end;
 end;
 
+procedure TInterfaceResolverInjected.DeleteResolver(aResolver: TInterfaceResolver);
+var
+  ix: integer;
+begin
+  ix := ObjArrayFind(fResolvers,aResolver);
+  if ix >= 0 then
+    ObjArrayDelete(fResolvers,ix);
```
Comment (synopse) [inline src/core/mormot.core.interfaces.pas:5401]: I guess you should also check within fResolversToBeReleased[] to avoid potential GPF in Destroy

Comment (synopse) [reply src/core/mormot.core.interfaces.pas:5401]: BTW I guess you could just write ObjArrayDelete(fResolvers, aResolver); with no intermediate ix variable

File: src/rest/mormot.rest.client.pas [hunk 4]
```diff
@@ -748,6 +748,9 @@   TRestClientUri = class(TRest)
     function ServiceDefineSharedApi(const aInterface: TGUID;
       const aContractExpected: RawUtf8 = SERVICE_CONTRACT_NONE_EXPECTED;
       aIgnoreAnyException: boolean = false): TServiceFactoryClient;
+    // undefine previously defined Interfaces
```
Comment (synopse) [inline src/rest/mormot.rest.client.pas:751]: please use 3 / for the first comment line ///
(for proper documentation generation)

File: src/rest/mormot.rest.client.pas [hunk 5]
```diff
@@ -748,6 +748,9 @@   TRestClientUri = class(TRest)
     function ServiceDefineSharedApi(const aInterface: TGUID;
       const aContractExpected: RawUtf8 = SERVICE_CONTRACT_NONE_EXPECTED;
       aIgnoreAnyException: boolean = false): TServiceFactoryClient;
+    // undefine previously defined Interfaces
+    procedure ServiceUndefine(const aInterfaces: array of TGUID);
+
```
Comment (synopse) [inline src/rest/mormot.rest.client.pas:753]: please don't add void lines

File: src/rest/mormot.rest.mvc.pas [hunk 6]
```diff
@@ -603,6 +603,9 @@   TMvcRunWithViews = class(TMvcRun)
   /// run TMvcApplication directly within a TRestServer method-based service
   // - this is the easiest way to host and publish a MVC Application, optionally
   // in conjunction with REST/AJAX client access
+
```
Comment (synopse) [inline src/rest/mormot.rest.mvc.pas:606]: please remove this comment generated by the IDE which breaks the main documentation comment

File: src/rest/mormot.rest.mvc.pas [hunk 7]
```diff
@@ -631,6 +635,7 @@   TMvcRunOnRestServer = class(TMvcRunWithViews)
       aViews: TMvcViewsAbstract = nil;
       aPublishOptions: TMvcPublishOptions=
         [low(TMvcPublishOption) .. high(TMvcPublishOption)]); reintroduce;
+    destructor Destroy; override;
```
Comment (synopse) [inline src/rest/mormot.rest.mvc.pas:638]: please add a short comment like
`/// finalize this MVC server logic instance`

File: src/rest/mormot.rest.mvc.pas [hunk 8]
```diff
@@ -2054,6 +2065,17 @@ procedure TMvcRunOnRestServer.RunOnRestServerSub(Ctxt: TRestServerUriContext);
     InternalRunOnRestServer(Ctxt, Ctxt.UriBlobFieldName);
 end;
 
+procedure TMvcRunOnRestServer.UnregisterMethods;
```
Comment (synopse) [inline src/rest/mormot.rest.mvc.pas:2068]: perhaps move it to TRestServer itself

File: src/rest/mormot.rest.server.pas [hunk 9]
```diff
@@ -6509,6 +6522,21 @@ procedure TRestServer.ServiceMethodRegister(aMethodName: RawUtf8;
   end;
 end;
 
+procedure TRestServer.ServiceMethodUnregister(aMethodName: RawUTF8);
+var
+  ix: Integer;
+begin
+  aMethodName := TrimU(aMethodName);
+  if aMethodName='' then
+    raise EServiceException.CreateUTF8('%.ServiceMethodRegister('''')',[self]);
```
Comment (synopse) [inline src/rest/mormot.rest.server.pas:6531]: please follow regular formating code with space after , := <>

File: src/rest/mormot.rest.server.pas [hunk 10]
```diff
@@ -6509,6 +6522,21 @@ procedure TRestServer.ServiceMethodRegister(aMethodName: RawUtf8;
...
 
+procedure TRestServer.ServiceMethodUnregister(aMethodName: RawUTF8);
+var
+  ix: Integer;
+begin
+  aMethodName := TrimU(aMethodName);
+  if aMethodName='' then
+    raise EServiceException.CreateUTF8('%.ServiceMethodRegister('''')',[self]);
+  ix:=fPublishedMethods.FindHashed(aMethodName);
+  if ix<>-1 then begin
```
Comment (synopse) [inline src/rest/mormot.rest.server.pas:6533]: hint: `if ix >=0 then` is safer and generates slightly better asm

File: src/soa/mormot.soa.core.pas [hunk 11]
```diff
@@ -689,6 +689,9 @@   TServiceContainer = class(TInterfaceResolverInjected)
     /// retrieve all registered Services contracts as a JSON array
     // - i.e. a JSON array of TServiceFactory.Contract JSON objects
     function AsJson: RawJson;
+    // delete previously registered Interfaces
```
Comment (synopse) [inline src/soa/mormot.soa.core.pas:692]: please use 3 / for the first comment line ///
(for proper documentation generation)

File: src/soa/mormot.soa.core.pas [hunk 12]
```diff
@@ -1590,7 +1593,35 @@ function TServiceContainer.CallBackUnRegister(const Callback: IInvokable): boole
...
+procedure TServiceContainer.DeleteInterface(const aInterfaces: PRttiInfoDynArray);
+var
+  iInterfaces: integer;
+  iMethods: integer;
+  Idx: integer;
+  MethodName: RawUTF8;
+  InterfaceName: RawUTF8;
+  tmpInterfaceName: RawUTF8;
+begin
+  for iInterfaces := Low(aInterfaces) to High(aInterfaces) do begin
```
Comment (synopse) [inline src/soa/mormot.soa.core.pas:1605]: please follow regular mORMot 2 code format

File: test/mormot2tests.lpi [hunk 13]
```diff
@@ -1,23 +1,24 @@
 <?xml version="1.0" encoding="UTF-8"?>
```
Comment (synopse) [inline test/mormot2tests.lpi:1]: please don't publish your own custom .lpi file

File: test/test.soa.core.pas [hunk 14]
```diff
@@ -1393,6 +1393,17 @@ procedure TTestServiceOrientedArchitecture.ServiceInitialization;
     'register TServiceCalculator as the ICalculator implementation on the server');
   // verify ICalculator RTTI-generated details
   Check(fClient.Server.Services <> nil);
+  if CheckFailed(fClient.Server.Services.Count = 1) then
```
Comment (synopse) [inline test/test.soa.core.pas:1396]: very nice seeing regression tests included!

**Review by synopse — CHANGES_REQUESTED** [review]
I commented some minor tweaks before merge.

Nice work!

File: src/core/mormot.core.interfaces.pas (hunk 2 above)
Comment (the-real-wizard-of-oz) [reply src/core/mormot.core.interfaces.pas:987]: Done

File: src/core/mormot.core.interfaces.pas (hunk 3 above)
Comment (the-real-wizard-of-oz) [reply src/core/mormot.core.interfaces.pas]: Done.

File: src/rest/mormot.rest.client.pas (hunk 4 above)
Comment (the-real-wizard-of-oz) [reply src/rest/mormot.rest.client.pas]: Done.

File: src/rest/mormot.rest.client.pas (hunk 5 above)
Comment (the-real-wizard-of-oz) [reply src/rest/mormot.rest.client.pas]: Done.

File: src/rest/mormot.rest.server.pas (hunk 10 above)
Comment (the-real-wizard-of-oz) [reply src/rest/mormot.rest.server.pas]: Done.

File: src/soa/mormot.soa.core.pas (hunk 11 above)
Comment (the-real-wizard-of-oz) [reply src/soa/mormot.soa.core.pas]: Done.

File: src/soa/mormot.soa.core.pas (hunk 12 above)
Comment (the-real-wizard-of-oz) [reply src/soa/mormot.soa.core.pas:1604]: Done.

File: test/mormot2tests.lpi (hunk 13 above)
Comment (the-real-wizard-of-oz) [reply test/mormot2tests.lpi]: Reverted .lpi file to master branch. Maybe we should add this file to git-ignore-file?

File: test/test.soa.core.pas (hunk 14 above)
Comment (the-real-wizard-of-oz) [reply test/test.soa.core.pas:1396]: ;)

File: src/rest/mormot.rest.mvc.pas (hunk 6 above)
Comment (the-real-wizard-of-oz) [reply src/rest/mormot.rest.mvc.pas]: Done.

File: src/rest/mormot.rest.server.pas (hunk 9 above)
Comment (the-real-wizard-of-oz) [reply src/rest/mormot.rest.server.pas]: Done.

File: src/rest/mormot.rest.mvc.pas (hunk 7 above)
Comment (the-real-wizard-of-oz) [reply src/rest/mormot.rest.mvc.pas:636]: Done.

File: src/rest/mormot.rest.mvc.pas (hunk 8 above)
Comment (the-real-wizard-of-oz) [reply src/rest/mormot.rest.mvc.pas:2066]: Does that make any sense? This code is for unregistering all methods introduced by the MVC interface. Moving to TRestServer would unregister ALL methods which is not what we want here. Imho this is right place.

**Comment by synopse** [comment]
Still some minor formatting issues, but I will fix them.

**Pull request merged**